Added "Useful Fact" effect

OnlyRealNubs opened the pull request:
![image](https://user-images.githubusercontent.com/91900600/161072679-b003015e-d068-482e-9a25-76b24414c158.png)
Shows a random fact from the list to display on the screen.

Add more, by simply adding more to the facts vector.

Reguas commented:
I guess it'd be better if these messages were shown as in-game help messages. Maybe take a look at [that](https://gtaforums.com/topic/820813-displaying-help-text/)?
 
![image](https://user-images.githubusercontent.com/64607261/161118943-35ca9a17-38b6-4e05-a250-3073222421fb.png)

OnlyRealNubs commented:
> I guess it'd be better if these messages were shown as in-game help messages

I can do that. Gimmie a minute.

OnlyRealNubs commented:
Done

Reguas commented:
Have you tested it? Cause `message` in `DISPLAY_HELP_TEXT_THIS_FRAME` seems to be a label that you have to set via `ADD_TEXT_ENTRY`.

OnlyRealNubs commented:
I have used that native before like that. I will check.

On Thu, 31 Mar 2022 at 2:55 pm Reguas ***@***.***> wrote:

> Have you tested it? Cause message in DISPLAY_HELP_TEXT_THIS_FRAME seems
> to be a label that you have to set via ADD_TEXT_ENTRY.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gta-chaos-mod/ChaosModV/pull/2808#issuecomment-1084985989>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV5EVOESH57UXQRZ5ROWB63VCXYJZANCNFSM5SFPNJZQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
364 Lancaster Squadron Cadet

Reguas commented on ChaosMod/Effects/db/Player/PlayerAdvice.cpp:
```diff
@@ -5,17 +5,34 @@
-	"If You Sprint, You Move Faster"
+	"If You Sprint, You Move Faster",
+	"~r~Th~o~is ~y~Is ~g~Rain~b~bow ~p~Text",
+	"This Game Was Made By Rockstar ~|~",
```
`¦` and `|` are different symbols lol

OnlyRealNubs replied:
I thought it was jut the font

OnlyRealNubs commented:
My Apologies, Very Tired Today

Reguas commented on ChaosMod/Effects/db/Player/PlayerAdvice.cpp:
```diff
@@ -38,7 +38,9 @@ std::vector<std::string> facts = {
+static void _GetMessage(int rng) //Without underscore, it's defined somewhere idek. I think it might be Microsoft
 {
 	msg = facts.at(rng);
 }
```
Why do you need this function?

Reguas commented on ChaosMod/Effects/db/Player/PlayerAdvice.cpp:
```diff
@@ -38,7 +38,9 @@ std::vector<std::string> facts = {
 int scaleform = 0;
 
-static void _GetMessage(int rng) //Without underscore, it's defined somewhere idek.
+int debugIndex = 0;
```
I guess, this shouldn't be here

OnlyRealNubs replied:
Thjat is when I was debugging it. One of the messages seemed to make the game crash.

Reguas commented on ChaosMod/Effects/db/Player/PlayerAdvice.cpp:
```diff
@@ -38,7 +38,9 @@ std::vector<std::string> facts = {
 std::string msg;
```
Why do you need this to be declared in global scope?

Reguas commented on ChaosMod/Effects/db/Player/PlayerAdvice.cpp:
```diff
@@ -0,0 +1,64 @@
+
+
+
+static void _GetMessage(int rng) //Without underscore, it's defined somewhere idek. I think it might be Microsoft
+{
+	msg = facts.at(rng);
+}
+
```
you still forgot to delete this

Reguas commented on ChaosMod/Effects/db/Player/PlayerAdvice.cpp:
```diff
@@ -0,0 +1,64 @@
+
+static void OnStart()
+{
+	std::string msg = facst.at(g_Random.GetRandomInt(0, facts.size()));
```
`facts` is misspelled

Reguas reviewed (changes requested):
Hey, are you making commits without building/testing? Not cool dude.

OnlyRealNubs commented:
My apologies. Very tired and busy. I will close this until this code is fixed and works.

OnlyRealNubs commented:
Alright. After doing more research on the display help text functions. I have remade the effect. 
Let's start from the top of the `OnStart` function

### Stuff I added and changed
- Now we check whether or not a help message is currently being displayed. If it is we remove it to show ours.
- We get a random message called `messageToDisplay` from the `messages` vector.
- I have completely scraped the while loop with the ticks, you will see why next.
- We set the help text, and then we display it. The function we display it with is called `END_TEXT_COMMAND_DISPLAY_HELP`. It takes 4 parameters. The first `p0` is always `0`, next is `loop` we set that to false as we do not want to loop the help message. Now `beep` set to `true` so it beeps. Finally we have `shape`, this is why the tick loop was scraped. If `shape` is set to a negative number it will display by default for 7.5 seconds. Anything above will just display it very briefly.

### What I learned
I learned to always research what you are doing, before you do it and create a pull request. And to make sure you test everything you do for bugs and other things.

Reguas commented:
That's great! Too much commentary for a simple effect, but you did learn important stuff.

OnlyRealNubs commented:
Forgot to mention. Do not spam this effect in the debug menu. It crashes the game. I don't believe it is the effect it self. But the game just not being able to keep, plus memory usage.

OnlyRealNubs commented on ChaosMod/Effects/db/Player/PlayerAdvice.cpp:
```diff
@@ -0,0 +1,56 @@
+	{
+		CLEAR_HELP(1);
+	}
+	std::string messageToDisplay = messages.at(g_Random.GetRandomInt(0, messages.size()));
```
Alright.
